In [1]:
# IMPORT LIBRARIES

# Pandas is used for reading and manipulating datasets
import pandas as pd

# NumPy helps with numerical computations
import numpy as np

# Machine Learning Algorithms
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LinearRegression

# Used to split dataset into training and testing data
from sklearn.model_selection import train_test_split

# Used to evaluate classification models
from sklearn.metrics import accuracy_score, confusion_matrix

# Converts text columns into numerical values
from sklearn.preprocessing import LabelEncoder

In [2]:
# Part A: Dataset Understanding

# 1. Load the dataset and display the first five records.

df = pd.read_csv('Dataset 2.csv')
df.head()

,UserID,Age,Gender,SubscriptionType,WatchHoursPerWeek,DevicesUsed,FavoriteGenre,AdClicks,MonthlySpend,SubscriptionRenewed
0,1001,22,Female,Basic,23,5,Comedy,13,353,No
1,1002,55,Male,Basic,9,4,Drama,14,317,Yes
2,1003,49,Male,Basic,8,3,Comedy,16,309,No
3,1004,39,Female,Premium,19,5,Drama,45,833,Yes
4,1005,38,Female,Premium,23,5,Sci-Fi,24,804,Yes


In [3]:
# 2. Determine the number of rows and columns in the dataset.

print('Rows and Columns:', df.shape)

Rows and Columns: (750, 10)


In [4]:
# 3. Display all column names.

print(df.columns)

Index(['UserID', 'Age', 'Gender', 'SubscriptionType', 'WatchHoursPerWeek',
       'DevicesUsed', 'FavoriteGenre', 'AdClicks', 'MonthlySpend',
       'SubscriptionRenewed'],
      dtype='object')


In [5]:
# 4. Identify numerical and categorical features.

numerical_features = df.select_dtypes(include=['int64','float64']).columns
categorical_features = df.select_dtypes(include=['object']).columns

print("Numerical Features:")
print(numerical_features)

print("\nCategorical Features:")
print(categorical_features)

Numerical Features:
Index(['UserID', 'Age', 'WatchHoursPerWeek', 'DevicesUsed', 'AdClicks',
       'MonthlySpend'],
      dtype='object')

Categorical Features:
Index(['Gender', 'SubscriptionType', 'FavoriteGenre', 'SubscriptionRenewed'], dtype='object')


In [6]:
# 5. Check whether the dataset contains missing values.

print(df.isnull().sum())

UserID                 0
Age                    0
Gender                 0
SubscriptionType       0
WatchHoursPerWeek      0
DevicesUsed            0
FavoriteGenre          0
AdClicks               0
MonthlySpend           0
SubscriptionRenewed    0
dtype: int64


In [7]:
# Part B: Exploratory Data Analysis

# 6. Calculate the average age of users.

print("Average Age:", df['Age'].mean())

Average Age: 41.824


In [8]:
# 7. Determine the average watch hours per week.

print("Average Watch Hours Per Week:", df['WatchHoursPerWeek'].mean())

Average Watch Hours Per Week: 14.236


In [9]:
# 8. Find the average monthly spending of users.

print("Average Monthly Spend:", df['MonthlySpend'].mean())

Average Monthly Spend: 689.9053333333334


In [10]:
# 9. Count the number of users in each subscription category.

print(df['SubscriptionType'].value_counts())

SubscriptionType
Basic      342
Premium    279
VIP        129
Name: count, dtype: int64


In [11]:
# 10. Determine the percentage of users who renewed their subscriptions.

renew_percentage = (df['SubscriptionRenewed'].value_counts(normalize=True)['Yes']) * 100
print("Renewal Percentage:", renew_percentage)

Renewal Percentage: 46.266666666666666


In [12]:
# Part C: Data Preparation

# 11. Convert categorical features into numerical form.

le = LabelEncoder()

for col in categorical_features:
    df[col] = le.fit_transform(df[col])

In [13]:
# 12. Define the feature set (X) and target variable (y) for subscription renewal prediction.

X = df.drop(['UserID','SubscriptionRenewed'], axis=1)
y = df['SubscriptionRenewed']

In [14]:
# 13. Split the dataset into training and testing sets.

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

In [15]:
# Part D: Decision Tree Classification

# 14. Train a Decision Tree model to predict whether a user will renew their subscription.

dt_model = DecisionTreeClassifier(random_state=42)
dt_model.fit(X_train, y_train)

DecisionTreeClassifier(random_state=42)

In [16]:
# 15. Evaluate the model using accuracy.

dt_pred = dt_model.predict(X_test)
dt_accuracy = accuracy_score(y_test, dt_pred)
print("Decision Tree Accuracy:", dt_accuracy)

Decision Tree Accuracy: 0.56


In [17]:
# 16. Generate and interpret the confusion matrix.

cm = confusion_matrix(y_test, dt_pred)
print("Confusion Matrix:")
print(cm)

Confusion Matrix:
[[48 34]
 [32 36]]


In [18]:
# Part E: K-Nearest Neighbors (KNN)

# 17. Train a KNN classifier with K = 5.

knn_model = KNeighborsClassifier(n_neighbors=5)
knn_model.fit(X_train, y_train)
knn_pred = knn_model.predict(X_test)

In [19]:
# 18. Compare the accuracy of KNN with the Decision Tree model.

knn_accuracy = accuracy_score(y_test, knn_pred)
print("KNN Accuracy:", knn_accuracy)
print("Decision Tree Accuracy:", dt_accuracy)

KNN Accuracy: 0.6333333333333333
Decision Tree Accuracy: 0.56


In [20]:
# Part F: Linear Regression

# 19. Train a Linear Regression model to predict monthly spending.

X_reg = df.drop(['UserID','MonthlySpend'], axis=1)
y_reg = df['MonthlySpend']

X_train_reg, X_test_reg, y_train_reg, y_test_reg = train_test_split(
    X_reg, y_reg, test_size=0.2, random_state=42
)

lr_model = LinearRegression()

lr_model.fit(X_train_reg, y_train_reg)

LinearRegression()

In [21]:
# 20. Predict the monthly spending for a new user and interpret the result.

new_user = [25,1,2,15,3,0,5,1]
prediction = lr_model.predict([new_user])

print("Predicted Monthly Spend:", prediction[0])

Predicted Monthly Spend: 1317.6691010329014


/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


Business Reflection Questions:-

1. Which factors appear to influence subscription renewal the most?

Based on the Decision Tree analysis, the most influential factors are:

MonthlySpend
AdClicks
Age
WatchHoursPerWeek
FavoriteGenre

Users who spend more, watch content regularly, and engage more with the platform are generally more likely to renew their subscriptions.

2. Why is subscription renewal a classification problem?

Subscription renewal is a classification problem because the target variable has two categories:

Yes (Renewed)
No (Not Renewed)

The model predicts which class a user belongs to rather than predicting a numerical value.


3. Why is monthly spending a regression problem?

Monthly spending is a regression problem because the target variable is a continuous numerical value (₹ amount).

For example:

₹350
₹750
₹1200

Since the output can take many numerical values, regression algorithms are used.


4. Which algorithm performed better for renewal prediction?

In this dataset:

Decision Tree Accuracy = 56.0%
KNN Accuracy = 63.3%

Therefore, K-Nearest Neighbors (KNN) performed better because it achieved higher prediction accuracy.


5. How could the platform use these predictions to improve customer retention?

Netflix can use these predictions to:

Identify users who are likely to cancel their subscriptions.
Offer personalized discounts and promotional plans.
Recommend movies and shows based on user preferences.
Increase engagement through targeted notifications and emails.
Improve customer satisfaction and reduce subscription churn.

By taking proactive actions for at-risk users, the platform can increase customer retention and revenue.